src/TorchSharp/Distributions/Constraints.cs

```diff
@@ -0,0 +1,637 @@
+                {
+                    public _IndependentConstraint(Constraint base_constraint, int reinterpreted_batch_ndims)
+                    {
+
+                    }
+
+                    public override Tensor check(Tensor value)
+                    {
+                        throw new NotImplementedException();
+                    }
+
+                    public override string ToString()
+                    {
+                        return this.GetType().Name + "()";
+                    }
+
+                    public Constraint base_constraint { get; private set; }
+
+                    /// <summary>
+                    /// Whether constrained space is discrete.
+                    /// </summary>
+                    public override bool is_discrete { get => base_constraint.is_discrete; }
+
+                    /// <summary>
+                    /// Number of rightmost dimensions that together define an event.
+                    /// The check() method will remove this many dimensions when computing validity.
+                    /// </summary>
+                    public override int event_dim { get => base_constraint.event_dim; }
+
+                }
+
+                /// <summary>
+                /// Constrain to the two values {0, 1}.
+                /// </summary>
+                public class _Boolean : Constraint
+                {
+                    public _Boolean() : base(true) { }
+
+                    public override Tensor check(Tensor value) => (value == 0) | (value == 1);
+                }
+
+                /// <summary>
+                /// Constrain to one-hot vectors.
+                /// </summary>
+                public class _OneHot : Constraint
+                {
+                    public _OneHot() : base(true, 1) { }
+
+                    public override Tensor check(Tensor value)
+                    {
+                        var is_boolean = (value == 0) | (value == 1);
+                        var is_normalized = value.sum(-1).eq(1);
+                        return is_boolean.all(-1) & is_normalized;
+                    }
+                }
+
+                /// <summary>
+                /// Constrain to an integer interval [lower_bound, upper_bound].
+                /// </summary>
+                public class _IntegerInterval : Constraint
+                {
+                    public _IntegerInterval(long lower_bound, long upper_bound) : base(true)
+                    {
+                        this.lower_bound = lower_bound;
+                        this.upper_bound = upper_bound;
+                    }
+
+                    public long lower_bound { get; private set; }
+
+                    public long upper_bound { get; private set; }
+
+                    public override Tensor check(Tensor value)
+                    {
+                        return (value % 1 == 0) & (lower_bound <= value) & (value <= upper_bound);
```

Comment:
>& [](http://example.com/codeflow?start=48&length=1)

I assume it is intentional to use the bit-operators here.

Reply:
Yes. Those are boolean tensors being ANDed together, not boolean expressions, where short-circuit operators would be appropriate.